In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn
from sklearn import preprocessing
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB

from sklearn.metrics import confusion_matrix,precision_score,accuracy_score
from sklearn.model_selection import GridSearchCV

In [3]:
nb_train = pd.read_csv("SalaryData_train.csv")
nb_test = pd.read_csv("SalaryData_test.csv")

In [4]:
nb_test.head()

age   workclass      education  educationno        maritalstatus  \
0   25     Private           11th            7        Never-married   
1   38     Private        HS-grad            9   Married-civ-spouse   
2   28   Local-gov     Assoc-acdm           12   Married-civ-spouse   
3   44     Private   Some-college           10   Married-civ-spouse   
4   34     Private           10th            6        Never-married   

           occupation    relationship    race    sex  capitalgain  \
0   Machine-op-inspct       Own-child   Black   Male            0   
1     Farming-fishing         Husband   White   Male            0   
2     Protective-serv         Husband   White   Male            0   
3   Machine-op-inspct         Husband   Black   Male         7688   
4       Other-service   Not-in-family   White   Male            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            50   United-States   <=50K  
2            0            40   United-States    >50K  
3            0            40   United-States    >50K  
4            0            30   United-States   <=50K

In [5]:
nb_test.tail()

age      workclass   education  educationno        maritalstatus  \
15055   33        Private   Bachelors           13        Never-married   
15056   39        Private   Bachelors           13             Divorced   
15057   38        Private   Bachelors           13   Married-civ-spouse   
15058   44        Private   Bachelors           13             Divorced   
15059   35   Self-emp-inc   Bachelors           13   Married-civ-spouse   

             occupation    relationship                 race      sex  \
15055    Prof-specialty       Own-child                White     Male   
15056    Prof-specialty   Not-in-family                White   Female   
15057    Prof-specialty         Husband                White     Male   
15058      Adm-clerical       Own-child   Asian-Pac-Islander     Male   
15059   Exec-managerial         Husband                White     Male   

       capitalgain  capitalloss  hoursperweek          native  Salary  
15055            0            0            40   United-States   <=50K  
15056            0            0            36   United-States   <=50K  
15057            0            0            50   United-States   <=50K  
15058         5455            0            40   United-States   <=50K  
15059            0            0            60   United-States    >50K

In [6]:
nb_train.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [7]:
nb_train.tail()

age      workclass    education  educationno        maritalstatus  \
30156   27        Private   Assoc-acdm           12   Married-civ-spouse   
30157   40        Private      HS-grad            9   Married-civ-spouse   
30158   58        Private      HS-grad            9              Widowed   
30159   22        Private      HS-grad            9        Never-married   
30160   52   Self-emp-inc      HS-grad            9   Married-civ-spouse   

               occupation relationship    race      sex  capitalgain  \
30156        Tech-support         Wife   White   Female            0   
30157   Machine-op-inspct      Husband   White     Male            0   
30158        Adm-clerical    Unmarried   White   Female            0   
30159        Adm-clerical    Own-child   White     Male            0   
30160     Exec-managerial         Wife   White   Female        15024   

       capitalloss  hoursperweek          native  Salary  
30156            0            38   United-States   <=50K  
30157            0            40   United-States    >50K  
30158            0            40   United-States   <=50K  
30159            0            20   United-States   <=50K  
30160            0            40   United-States    >50K

## EDA

In [8]:
nb_test.columns

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [9]:
nb_train.columns

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [10]:
nb_test.isnull().sum()

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
Salary           0
dtype: int64

In [11]:
nb_train.isnull().sum()

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
Salary           0
dtype: int64

In [12]:
nb_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15060 entries, 0 to 15059
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            15060 non-null  int64 
 1   workclass      15060 non-null  object
 2   education      15060 non-null  object
 3   educationno    15060 non-null  int64 
 4   maritalstatus  15060 non-null  object
 5   occupation     15060 non-null  object
 6   relationship   15060 non-null  object
 7   race           15060 non-null  object
 8   sex            15060 non-null  object
 9   capitalgain    15060 non-null  int64 
 10  capitalloss    15060 non-null  int64 
 11  hoursperweek   15060 non-null  int64 
 12  native         15060 non-null  object
 13  Salary         15060 non-null  object
dtypes: int64(5), object(9)
memory usage: 1.6+ MB


In [13]:
nb_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


In [14]:
nb_train.dtypes

age               int64
workclass        object
education        object
educationno       int64
maritalstatus    object
occupation       object
relationship     object
race             object
sex              object
capitalgain       int64
capitalloss       int64
hoursperweek      int64
native           object
Salary           object
dtype: object

In [19]:
nb_test.dtypes

age                 int64
workclass        category
education        category
educationno         int64
maritalstatus    category
occupation       category
relationship     category
race             category
sex              category
capitalgain         int64
capitalloss         int64
hoursperweek        int64
native           category
Salary           category
dtype: object

In [16]:
nb_train['workclass']=nb_train['workclass'].astype('category')
nb_train['education']=nb_train['education'].astype('category')
nb_train['maritalstatus']=nb_train['maritalstatus'].astype('category')
nb_train['occupation']=nb_train['occupation'].astype('category')
nb_train['relationship']=nb_train['relationship'].astype('category')
nb_train['race']=nb_train['race'].astype('category')
nb_train['sex']=nb_train['sex'].astype('category')
nb_train['native']=nb_train['native'].astype('category')
nb_train['Salary']=nb_train['Salary'].astype('category')

In [18]:
nb_test['workclass']=nb_test['workclass'].astype('category')
nb_test['education']=nb_test['education'].astype('category')
nb_test['maritalstatus']=nb_test['maritalstatus'].astype('category')
nb_test['occupation']=nb_test['occupation'].astype('category')
nb_test['relationship']=nb_test['relationship'].astype('category')
nb_test['race']=nb_test['race'].astype('category')
nb_test['sex']=nb_test['sex'].astype('category')
nb_test['native']=nb_test['native'].astype('category')
nb_test['Salary']=nb_test['Salary'].astype('category')

In [23]:
label_encoder=preprocessing.LabelEncoder()
nb_train['workclass']=label_encoder.fit_transform(nb_train['workclass'])
nb_train['education']=label_encoder.fit_transform(nb_train['education'])
nb_train['maritalstatus']=label_encoder.fit_transform(nb_train['maritalstatus'])
nb_train['occupation']=label_encoder.fit_transform(nb_train['occupation'])
nb_train['relationship']=label_encoder.fit_transform(nb_train['relationship'])
nb_train['race']=label_encoder.fit_transform(nb_train['race'])
nb_train['sex']=label_encoder.fit_transform(nb_train['sex'])
nb_train['native']=label_encoder.fit_transform(nb_train['native'])
nb_train['Salary']=label_encoder.fit_transform(nb_train['Salary'])


In [24]:
nb_test['workclass']=label_encoder.fit_transform(nb_test['workclass'])
nb_test['education']=label_encoder.fit_transform(nb_test['education'])
nb_test['maritalstatus']=label_encoder.fit_transform(nb_test['maritalstatus'])
nb_test['occupation']=label_encoder.fit_transform(nb_test['occupation'])
nb_test['relationship']=label_encoder.fit_transform(nb_test['relationship'])
nb_test['race']=label_encoder.fit_transform(nb_test['race'])
nb_test['sex']=label_encoder.fit_transform(nb_test['sex'])
nb_test['native']=label_encoder.fit_transform(nb_test['native'])
nb_test['Salary']=label_encoder.fit_transform(nb_test['Salary'])


In [25]:
nb_train.head()

age  workclass  education  educationno  maritalstatus  occupation  \
0   39          5          9           13              4           0   
1   50          4          9           13              2           3   
2   38          2         11            9              0           5   
3   53          2          1            7              2           5   
4   28          2          9           13              2           9   

   relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  \
0             1     4    1         2174            0            40      37   
1             0     4    1            0            0            13      37   
2             1     4    1            0            0            40      37   
3             0     2    1            0            0            40      37   
4             5     2    0            0            0            40       4   

   Salary  
0       0  
1       0  
2       0  
3       0  
4       0

In [26]:
nb_train.head(10)

age  workclass  education  educationno  maritalstatus  occupation  \
0   39          5          9           13              4           0   
1   50          4          9           13              2           3   
2   38          2         11            9              0           5   
3   53          2          1            7              2           5   
4   28          2          9           13              2           9   
5   37          2         12           14              2           3   
6   49          2          6            5              3           7   
7   52          4         11            9              2           3   
8   31          2         12           14              4           9   
9   42          2          9           13              2           3   

   relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  \
0             1     4    1         2174            0            40      37   
1             0     4    1            0            0            13      37   
2             1     4    1            0            0            40      37   
3             0     2    1            0            0            40      37   
4             5     2    0            0            0            40       4   
5             5     4    0            0            0            40      37   
6             1     2    0            0            0            16      21   
7             0     4    1            0            0            45      37   
8             1     4    0        14084            0            50      37   
9             0     4    1         5178            0            40      37   

   Salary  
0       0  
1       0  
2       0  
3       0  
4       0  
5       0  
6       0  
7       1  
8       1  
9       1

In [27]:
nb_test.head()

age  workclass  education  educationno  maritalstatus  occupation  \
0   25          5          9            7              4           0   
1   38          4          9            9              2           3   
2   28          2         11           12              0           5   
3   44          2          1           10              2           5   
4   34          2          9            6              2           9   

   relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  \
0             1     4    1            0            0            40      37   
1             0     4    1            0            0            50      37   
2             1     4    1            0            0            40      37   
3             0     2    1         7688            0            40      37   
4             5     2    0            0            0            30       4   

   Salary  
0       0  
1       0  
2       0  
3       0  
4       0

In [28]:
nb_test.head(10)

age  workclass  education  educationno  maritalstatus  occupation  \
0   25          5          9            7              4           0   
1   38          4          9            9              2           3   
2   28          2         11           12              0           5   
3   44          2          1           10              2           5   
4   34          2          9            6              2           9   
5   63          2         12           15              2           3   
6   24          2          6           10              3           7   
7   55          4         11            4              2           3   
8   65          2         12            9              4           9   
9   36          2          9           13              2           3   

   relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  \
0             1     4    1            0            0            40      37   
1             0     4    1            0            0            50      37   
2             1     4    1            0            0            40      37   
3             0     2    1         7688            0            40      37   
4             5     2    0            0            0            30       4   
5             5     4    0         3103            0            32      37   
6             1     2    0            0            0            40      21   
7             0     4    1            0            0            10      37   
8             1     4    0         6418            0            40      37   
9             0     4    1            0            0            40      37   

   Salary  
0       0  
1       0  
2       0  
3       0  
4       0  
5       0  
6       0  
7       1  
8       1  
9       1

In [29]:
X_train=nb_train.iloc[0:,0:13]

In [30]:
X_train

age  workclass  education  educationno  maritalstatus  occupation  \
0       39          5          9           13              4           0   
1       50          4          9           13              2           3   
2       38          2         11            9              0           5   
3       53          2          1            7              2           5   
4       28          2          9           13              2           9   
...    ...        ...        ...          ...            ...         ...   
30156   27          2          7           12              2          12   
30157   40          2         11            9              2           6   
30158   58          2         11            9              6           0   
30159   22          2         11            9              4           0   
30160   52          3         11            9              2           3   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  
0                 1     4    1         2174            0            40      37  
1                 0     4    1            0            0            13      37  
2                 1     4    1            0            0            40      37  
3                 0     2    1            0            0            40      37  
4                 5     2    0            0            0            40       4  
...             ...   ...  ...          ...          ...           ...     ...  
30156             5     4    0            0            0            38      37  
30157             0     4    1            0            0            40      37  
30158             4     4    0            0            0            40      37  
30159             3     4    1            0            0            20      37  
30160             5     4    0        15024            0            40      37  

[30161 rows x 13 columns]

In [31]:
y_train=nb_train.iloc[0:,-1]
y_train

0        0
1        0
2        0
3        0
4        0
        ..
30156    0
30157    1
30158    0
30159    0
30160    1
Name: Salary, Length: 30161, dtype: int32

In [32]:
X_test=nb_test.iloc[0:,0:13]

In [33]:
X_test

age  workclass  education  educationno  maritalstatus  occupation  \
0       25          5          9            7              4           0   
1       38          4          9            9              2           3   
2       28          2         11           12              0           5   
3       44          2          1           10              2           5   
4       34          2          9            6              2           9   
...    ...        ...        ...          ...            ...         ...   
15055   33          2         11           13              5           0   
15056   39          4          9           13              2           2   
15057   38          2          9           13              2           3   
15058   44          1         15           13              4          10   
15059   35          2         11           13              2          11   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  
0                 1     4    1            0            0            40      37  
1                 0     4    1            0            0            50      37  
2                 1     4    1            0            0            40      37  
3                 0     2    1         7688            0            40      37  
4                 5     2    0            0            0            30       4  
...             ...   ...  ...          ...          ...           ...     ...  
15055             1     4    1            0            0            40       3  
15056             0     4    1            0            0            36      37  
15057             0     4    1            0            0            50      37  
15058             3     4    1         5455            0            40      37  
15059             0     4    1            0            0            60      37  

[15060 rows x 13 columns]

In [34]:
y_test=nb_test.iloc[0:,-1]

In [35]:
y_test

0        0
1        0
2        0
3        0
4        0
        ..
15055    0
15056    1
15057    1
15058    0
15059    0
Name: Salary, Length: 15060, dtype: int32

In [36]:
gb_n=GaussianNB()
gb_n.fit(X_train,y_train)

GaussianNB()

### Checking train accuracy

In [37]:
y_pred_train = gb_n.predict(X_train)

In [38]:
y_pred_df=pd.DataFrame({'actual_train':y_train, 'predicted_preb':gb_n.predict(X_train)})

In [39]:
y_pred_df

actual_train  predicted_preb
0                 0               0
1                 0               0
2                 0               0
3                 0               0
4                 0               0
...             ...             ...
30156             0               0
30157             1               0
30158             0               0
30159             0               0
30160             1               1

[30161 rows x 2 columns]

In [40]:
confusion_matrix(y_train,y_pred_train)

array([[21505,  1148],
       [ 5025,  2483]], dtype=int64)

In [41]:
accuracy_score(y_train,y_pred_train)

0.7953317197705646

In [42]:
precision_score(y_train,y_pred_train)

0.6838336546405949

### Checking test accuracy

In [43]:
y_pred=gb_n.predict(X_test)

In [45]:
confusion_matrix(y_test,y_pred)

array([[10050,  1290],
       [ 3287,   433]], dtype=int64)

In [46]:
y_pred_df=pd.DataFrame({'actual_train': y_test, 'predicted_preb':gb_n.predict(X_test)})

In [47]:
y_pred_df

actual_train  predicted_preb
0                 0               0
1                 0               0
2                 0               0
3                 0               1
4                 0               0
...             ...             ...
15055             0               0
15056             1               0
15057             1               0
15058             0               1
15059             0               0

[15060 rows x 2 columns]

In [48]:
precision_score(y_test,y_pred)

0.2513058618688334

In [49]:
accuracy_score(y_test,y_pred)

0.6960823373173971

In [51]:
mb_n = MultinomialNB()
mb_n.fit(X_train,y_train)

MultinomialNB()

In [52]:
y_pred_Train_Mb=mb_n.predict(X_train)

### checking accuracy for training

In [53]:
y_pred_DF1 = pd.DataFrame({'actual_train': y_train, 'predicted_train':mb_n.predict(X_train)})

In [54]:
y_pred_DF1

actual_train  predicted_train
0                 0                1
1                 0                0
2                 0                0
3                 0                0
4                 0                0
...             ...              ...
30156             0                0
30157             1                0
30158             0                0
30159             0                0
30160             1                1

[30161 rows x 2 columns]

In [55]:
confusion_matrix(y_train,mb_n.predict(X_train))

array([[21717,   936],
       [ 5913,  1595]], dtype=int64)

In [56]:
accuracy_score(y_train,mb_n.predict(X_train))

0.7729186698053778

In [57]:
precision_score(y_train,mb_n.predict(X_train))

0.6301856973528249

### Checking accuracy for testing

In [58]:
y_pred = mb_n.predict(X_test)

In [59]:
y_pred_df = pd.DataFrame({'actual': y_test, 'predicted' : mb_n.predict(X_test)})

In [60]:
y_pred_df

actual  predicted
0           0          0
1           0          0
2           0          0
3           0          1
4           0          0
...       ...        ...
15055       0          0
15056       1          0
15057       1          0
15058       0          1
15059       0          0

[15060 rows x 2 columns]

In [61]:
confusion_matrix(y_test,y_pred)

array([[10385,   955],
       [ 3426,   294]], dtype=int64)

In [62]:
precision_score(y_test,y_pred)

0.2353883106485188

In [64]:
accuracy_score(y_test,y_pred)

0.7090969455511288

### with multinomial naivebayes the accuracy score is 0.709, For multinomial naive bayes the training and testing accuracy is same